ssl/ssl_lib.c

```diff
@@ -3316,6 +3316,12 @@ int SSL_do_handshake(SSL *s)
 
     s->method->ssl_renegotiate_check(s, 0);
 
+    if (SSL_is_server(s)) {
+        /* clear SNI settings at server-side */
+        OPENSSL_free(s->ext.hostname);
+        s->ext.hostname = NULL;
```

Comment:
The idea here is clearing SNI data at 'operation'/'run-time' phase instead of at configuration phase. One benefit of doing this is that the 'side' information is already fixed at `SSL_do_handshake`, and `SSL_is_server` will *always* get the correct side info (hope I don't miss something...). `SSL_set_accept_state` is still the 'configuration',  and the 'contrived example' example you mentioned I think it's just an example of supporting clearing SNI info at `SSL_do_handshake`, that is: whatever ways a user switches sides (by using set_accept_state and set_connect_state), we are not bothered, just do the clearance when the user wants to do some real jobs...